kernel/src/hil/flash.rs

```diff
@@ -4,141 +4,150 @@
-//! }
 //! ```
 //!
 //! A user of this flash interface might look like:
 //!
 //! ```rust
+//! #![feature(min_const_generics)]
 //! use kernel::common::cells::TakeCell;
-//! use kernel::hil;
+//! use kernel::hil::flash;
+//! use kernel::ReturnCode;
 //!
-//! pub struct FlashUser<'a, F: hil::flash::Flash + 'static> {
+//! pub struct FlashUser<'a, F: flash::Flash<W, E> + 'static, const W: usize, const E: usize> {
 //!     driver: &'a F,
-//!     buffer: TakeCell<'static, F::Page>,
+//!     buffer: TakeCell<'static, [u8; W]>,
 //! }
 //!
-//! impl<'a, F: hil::flash::Flash> FlashUser<'a, F> {
-//!     pub fn new(driver: &'a F, buffer: &'static mut F::Page) -> FlashUser<'a, F> {
+//! impl<'a, F: flash::Flash<W, E>, const W: usize, const E: usize> FlashUser<'a, F, W, E> {
+//!     pub fn new(driver: &'a F, buffer: &'static mut [u8; W]) -> FlashUser<'a, F, W, E> {
 //!         FlashUser {
 //!             driver: driver,
 //!             buffer: TakeCell::new(buffer),
 //!         }
 //!     }
 //! }
 //!
-//! impl<'a, F: hil::flash::Flash> hil::flash::Client<F> for FlashUser<'a, F> {
-//!     fn read_complete(&self, buffer: &'static mut F::Page, error: hil::flash::Error) {}
-//!     fn write_complete(&self, buffer: &'static mut F::Page, error: hil::flash::Error) { }
-//!     fn erase_complete(&self, error: hil::flash::Error) {}
+//! impl<'a, F: flash::Flash<W, E>, const W: usize, const E: usize> flash::Client<W, E> for FlashUser<'a, F, W, E> {
+//!     fn read_complete(&self, read_buffer: &'static mut [u8], ret: Result<(), ReturnCode>) {}
+//!     fn write_complete(&self, write_buffer: &'static mut [u8; W], ret: Result<(), ReturnCode>) { }
+//!     fn erase_complete(&self, ret: Result<(), ReturnCode>) {}
 //! }
 //! ```
 
 use crate::returncode::ReturnCode;
 
-/// Flash errors returned in the callbacks.
-#[derive(Copy, Clone, Debug, Eq, PartialEq)]
-pub enum Error {
-    /// Success.
-    CommandComplete,
-
-    /// An error occurred during the flash operation.
-    FlashError,
-}
-
 pub trait HasClient<'a, C> {
     /// Set the client for this flash peripheral. The client will be called
     /// when operations complete.
     fn set_client(&'a self, client: &'a C);
 }
 
-/// A page of writable persistent flash memory.
-pub trait Flash {
-    /// Type of a single flash page for the given implementation.
-    type Page: AsMut<[u8]> + Default;
-
-    /// Read a page of flash into the buffer.
-    fn read_page(
+/// A page of writeable persistent flash memory.
+///
+/// `W`: Should be the minimum number of bytes that can be written
+///      in an operation.
+/// `E`: Should be the minimum number of bytes that can be erased
+///      in an operation.
+pub trait Flash<const W: usize, const E: usize> {
+    /// Read data from flash into a buffer.
+    ///
+    /// This function will read data stored in flash at `address` and
+    /// `length` into the buffer `buf`.
+    ///
+    /// Not all implementations support this, in that case `read_page`
```

Comment:
Was `read_page` removed intentionally? Unless I'm misreading this diff, it looks gone?

Reply:
Yes, it has just been replaced with a `read()` function.